src/devtools/views/Components/InspectedElementContext.js

```diff
@@ -0,0 +1,176 @@
+const resource: Resource<number, number, InspectedElement> = createResource(
+  (id: number) => {
+    let request = inProgressRequests.get(id);
+    if (request != null) {
+      return request.promise;
+    }
+
+    let resolveFn = ((null: any): ResolveFn);
+    const promise = new Promise(resolve => {
+      resolveFn = resolve;
+    });
+
+    inProgressRequests.set(id, { promise, resolveFn });
+
+    return promise;
+  },
+  (id: number) => id,
+  { useLRU: true }
+);
+
+type Props = {|
+  children: React$Node,
+|};
+
+function InspectedElementContextController({ children }: Props) {
+  const bridge = useContext(BridgeContext);
+  const store = useContext(StoreContext);
+
+  // It's very important that this context consumes selectedElementID and not inspectedElementID.
+  // Otherwise the effect that sends the "inspect" message across the bridge-
+  // would itself be blocked by the same render that suspends (waiting for the data).
+  const { selectedElementID } = useContext(TreeStateContext);
+
+  const [count, setCount] = useState<number>(0);
+
+  // This effect handler invalidates the suspense cache and schedules rendering updates with React.
+  useEffect(() => {
+    const onInspectedElement = (
+      inspectedElementResponse: InspectedElementResponse | null
+    ) => {
+      if (inspectedElementResponse != null) {
+        let { inspectedElement } = inspectedElementResponse;
+        if (inspectedElement !== null) {
+          const id = inspectedElement.id;
+
+          inspectedElement = (({
+            ...inspectedElement,
+            context: hydrateHelper(inspectedElement.context),
+            hooks: hydrateHelper(inspectedElement.hooks),
+            props: hydrateHelper(inspectedElement.props),
+            state: hydrateHelper(inspectedElement.state),
+          }: any): InspectedElement);
+
+          const request = inProgressRequests.get(id);
+          if (request != null) {
+            inProgressRequests.delete(id);
+            request.resolveFn(inspectedElement);
+          } else {
+            resource.write(id, inspectedElement);
+
+            // Schedule update with React if the curently-selected element has been invalidated.
+            if (id === selectedElementID) {
+              setCount(count => count + 1);
+            }
+          }
+        }
+      }
+    };
+
+    bridge.addListener('inspectedElement', onInspectedElement);
+    return () => bridge.removeListener('inspectedElement', onInspectedElement);
+  }, [bridge, selectedElementID]);
+
+  // This effect handler polls for updates on the currently selected element.
+  useEffect(() => {
+    if (selectedElementID === null) {
+      return () => {};
+    }
+
+    const rendererID = store.getRendererIDForElement(selectedElementID);
+
+    let timeoutID: TimeoutID | null = null;
+
+    const sendRequest = () => {
+      timeoutID = null;
+
+      bridge.send('inspectElement', { id: selectedElementID, rendererID });
+    };
+
```

Comment:
Might want to add

```js
bridge.send('selectElement', { id: selectedElementID, rendererID });
```

here to update `$r`.

Reply:
👍